### PR TITLE
feat: add support for impersonation

### LINKED ISSUE #417: Support for --impersonate-service-account
It would be nice, if Cloud SQL Proxy supports the `--impersonate-service-account` flag similar to the corresponding flag in `gcloud`. This would help to use impersonation out-of-the-box instead of long-running and possibly non-personalized service-account credentials.

### PR DESCRIPTION
**Pull request by enocom**
Note: the CLI flag matches the gcloud UI.

Fixes #417

### CODE REVIEW

**Comment by enocom** [comment]
~361 lines of this PR are just cleaning up our e2e authentication tests to be table tests, and adding a matching impersonation test for each authentication type.

**Comment by enocom** [comment]
Also, I've updated our CI pipeline to support impersonation as well.

**Review by hessjcg — APPROVED** [review]
LGTM

File: cmd/root.go [hunk 1]
```diff
@@ -347,6 +360,16 @@ func parseConfig(cmd *Command, conf *proxy.Config, args []string) error {
 		}
 	}
 
+	if cmd.impersonationChain != "" {
+		accts := strings.Split(cmd.impersonationChain, ",")
```
Comment (kurtisvg) [inline cmd/root.go:364]: Can this ever be an empty slice?

File: cmd/root.go [hunk 2]
```diff
@@ -347,6 +360,16 @@ func parseConfig(cmd *Command, conf *proxy.Config, args []string) error {
...
+		conf.ImpersonateTarget = accts[0]
+		// Assign delegates if the chain is more than one account.
+		if l := len(accts); l > 1 {
+			conf.ImpersonateTarget = accts[l-1]
+			conf.ImpersonateDelegates = accts[:l-1]
+		}
```
Comment (kurtisvg) [inline cmd/root.go:370]: nit:
```suggestion
                 l := len(accts)
		conf.ImpersonateTarget = accts[l-1]
		// Assign delegates if the chain is more than one account.
		if l > 1 {
			conf.ImpersonateDelegates = accts[:l-1]
		}
```

File: internal/proxy/proxy.go [hunk 3]
```diff
@@ -187,38 +199,89 @@ func (c *Config) DialOptions(i InstanceConnConfig) []cloudsqlconn.DialOption {
...
+		default:
+			l.Infof("Impersonating service account with Application Default Credentials")
+		}
+		ts, err := impersonate.CredentialsTokenSource(
+			context.Background(),
+			impersonate.CredentialsConfig{
+				TargetPrincipal: c.ImpersonateTarget,
+				Delegates:       c.ImpersonateDelegates,
+				Scopes: []string{
+					sqladmin.CloudPlatformScope,
```
Comment (kurtisvg) [inline internal/proxy/proxy.go:235]: Do we need CloudPlatform here?

File: internal/proxy/proxy.go [hunk 4]
```diff
@@ -187,38 +199,89 @@ func (c *Config) DialOptions(i InstanceConnConfig) []cloudsqlconn.DialOption {
 	return opts
 }
 
-// DialerOptions builds appropriate list of options from the Config
-// values for use by cloudsqlconn.NewClient()
-func (c *Config) DialerOptions(l cloudsql.Logger) ([]cloudsqlconn.Option, error) {
-	opts := []cloudsqlconn.Option{
-		cloudsqlconn.WithUserAgent(c.UserAgent),
+func (c *Config) credentialsOpt(l cloudsql.Logger) (cloudsqlconn.Option, error) {
```
Comment (kurtisvg) [inline internal/proxy/proxy.go:202]: This doesn't look like it modifies c in anyway. Would it be better to take c as an arg instead?

**Review by kurtisvg — CHANGES_REQUESTED** [review]

File: cmd/root.go [hunk 5]
```diff
@@ -253,6 +260,9 @@ https://cloud.google.com/storage/docs/requester-pays`)
 	cmd.PersistentFlags().StringVar(&c.conf.FUSETempDir, "fuse-tmp-dir",
 		filepath.Join(os.TempDir(), "csql-tmp"),
 		"Temp dir for Unix sockets created with FUSE")
+	cmd.PersistentFlags().StringVar(&c.impersonationChain, "impersonate-service-account", "",
+		`Comma separated list of service accounts to impersonate. Last value
+is the target account.`)
```
[the remark this answers is not in the document]
Comment (enocom) [reply cmd/root.go:265]: Good question. I checked the gcloud docs and the Impersonate docs again. Both mention service accounts only and not IAM identities. Cf. https://pkg.go.dev/google.golang.org/api/impersonate#CredentialsConfig.

File: cmd/root.go (hunk 1 above)
Comment (enocom) [reply cmd/root.go:364]: Only if a user passed this: `--impersonate-service-account=,`.

File: internal/proxy/proxy.go (hunk 3 above)
Comment (enocom) [reply internal/proxy/proxy.go:235]: Let me check on that -- we did for the e2e tests using the token.

File: internal/proxy/proxy.go (hunk 4 above)
Comment (enocom) [reply internal/proxy/proxy.go:202]: Yes, I'm in favor of that idea.

Comment (enocom) [reply internal/proxy/proxy.go:202]: Looking at this again, I see we're reading a bunch of data from `c`. So even though we're not modifying it, it's still common to define `c` as a receiver. Passing it as an argument would basically be a de-sugared version of the same code here.

File: cmd/root.go (hunk 2 above)
Comment (enocom) [reply cmd/root.go]: Done.

File: internal/proxy/proxy.go (hunk 3 above)
Comment (enocom) [reply internal/proxy/proxy.go]: We don't need it here. It's the token source that gets the impersonated token that needs this scope. Removed it here, and removed the SQL Admin scope from the token source in the tests.

**Comment by enocom** [comment]
I need to verify the delegation logic here. I've matched gcloud's behavior, but need to check if the Go library does the same (back to front).

File: internal/proxy/proxy.go (hunk 4 above)
Comment (kurtisvg) [reply internal/proxy/proxy.go:202]: downgrade to nit: but it does seem like a receiver indicates we are doing something to `c`, where a read only parameter makes clear in the intent is only to create something from c.

File: tests/connection_test.go [hunk 6]
```diff
@@ -35,7 +34,9 @@ const connTestTimeout = time.Minute
 // and then unsets GOOGLE_APPLICATION_CREDENTIALS. It returns a cleanup function
 // that restores the original setup.
 func removeAuthEnvVar(t *testing.T) (*oauth2.Token, string, func()) {
-	ts, err := google.DefaultTokenSource(context.Background(), sqladmin.SqlserviceAdminScope)
+	ts, err := google.DefaultTokenSource(context.Background(),
+		"https://www.googleapis.com/auth/cloud-platform",
```
Comment (kurtisvg) [inline tests/connection_test.go:38]: Why do we need cloud-platform here?

**Review by kurtisvg — APPROVED** [review]

File: internal/proxy/proxy.go (hunk 4 above)
Comment (enocom) [reply internal/proxy/proxy.go:202]: Let me try that in a separate PR and we can apply the pattern across the board if we like it.

File: tests/connection_test.go (hunk 6 above)
Comment (enocom) [reply tests/connection_test.go:38]: Short answer: the impersonation fails without this scope. Longer answer: let me figure out if we can limit this scope still further.

Comment (enocom) [reply tests/connection_test.go:38]: After some digging, I found that this is the required OAuth2 scope to call generateAccessToken.


See:
- https://cloud.google.com/iam/docs/reference/credentials/rest/v1/projects.serviceAccounts/generateAccessToken
- https://github.com/googleapis/google-api-go-client/blob/main/iamcredentials/v1/iamcredentials-gen.go#L85

File: internal/proxy/proxy.go (hunk 4 above)
Comment (enocom) [reply internal/proxy/proxy.go:202]: https://github.com/GoogleCloudPlatform/cloud-sql-proxy/pull/1480

**Pull request merged**